tutorials/getstarted.qmd

```diff
@@ -0,0 +1,471 @@
+
+## Allocating and in-place computations
+
+Memory allocation is a [critical performace issue](https://docs.julialang.org/en/v1/manual/performance-tips/#Measure-performance-with-[@time](@ref)-and-pay-attention-to-memory-allocation) when programming in Julia. To take this into account, `Manifolds.jl` provides special functions to reduce the amount of allocations.
+
+We again look at the [📖 exponential map](https://en.wikip edia.org/wiki/Exponential_map_(Riemannian_geometry)). On a manifold `M` the exponential map needs a point `p` (to start from) and a tangent vector `X`, which can be seen as direction to “walk into” as well as the length to walk into this direction. In `Manifolds.jl` the function can then be called with `q = exp(M, p, X)` [🔗](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/functions.html#Base.exp-Tuple{AbstractManifold,%20Any,%20Any}). This function returns the resulting point `q`, which requires to allocate new memory.
+
+To avoid this allocation, the function `exp!(M, q, p, X)` [🔗](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/functions.html#ManifoldsBase.exp!-Tuple{AbstractManifold,%20Any,%20Any,%20Any}) can be called. Here `q` is allocated beforehand and is passed as the memory, where the result is returned in.
+It might be used even for interims computations, as long as it does not introduce side effects. Thas means that even with `exp!(M, p, p, X)` the result is correct.
+
+Let's look at an example.
+
+We take another look at the [`Sphere`](https://juliamanifolds.github.io/Manifolds.jl/latest/manifolds/sphere.html#Manifolds.Sphere) but now a high-dimensional one.
+We can also illustrate how to generate radnom points and tangent vectors.
+
+```{julia}
+#| output: false
+M = Sphere(10000)
+p₄ = rand(M)
+X = rand(M; vector_at=p₄)
+```
+
+Looking at the allocations required we get
+
+```{julia}
+@allocated exp(M, p₄, X)
+```
+
+While if we have already allocated memory for the resulting point on the manifold,
+for example
+
+```{julia}
+q₂ = zero(p₄);
+```
+
+There are no new memory allocations necessary if we use the in-place function."
+
+```{julia}
+@allocated exp!(M, q₂, p₄, X)
+```
+
+
+This methodology is used for all functions that compute a new point or tangent vector. By default all allocating functions allocate memory and call the in-place function.
+This also means that if you implement a new manifold, you just have to implement the in-place version.
+
+## Decorating a manifold"
+
+As you saw until now, an
+[📎 `AbstractManifold`](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/types.html#The-AbstractManifold)
```

Comment:
Why is there a paperclip here?

Reply:
they should all have because it is links to outside the Manifolds.jl docs :) they all should have that icon. But now I am back at my computer and can fix those.

Reply:
Ah missed that one and misread it maybe, its fixed now (I think in a very early version paperclip was meant to be  🔗)